### PR TITLE
Allow Page to show actions on the start or end of the title bar

### PR DESCRIPTION
**Pull request by mjopson**
## Summary

Hi! Thought I'd jump right in with this PR...

Allows setting the position of actions as described in https://github.com/contiamo/operational-ui/issues/676.  This doesn't address the idea of having buttons with configurable icon positions; figured that was a separate change.

You do get a bit of extra spacing between elements when putting actions on the left; not sure if this is desired or not.

image

## Related issue

https://github.com/contiamo/operational-ui/issues/676

## To be tested

Me
- [x] No error/warning in the console

Tester 1

- [ ] The netlify build is working
  

Tester 2

- [ ] The netlify build is working

### CODE REVIEW

**Comment by TejasQ** [comment]
Thanks so much for your contribution, @mjopson! Great stuff! Let's quickly check with our core designer about the extra spacing on the left. @micha-f, how's it looking?

File: src/Page/Page.tsx [hunk 1]
```diff
@@ -104,7 +105,7 @@ const ViewContainer = styled("div")<{ isInTab?: boolean }>(({ theme, isInTab })
   position: "relative",
 }))
 
-const ActionsContainer = styled("div")(({ theme }) => ({
+const EndContainer = styled("div")(({ theme }) => ({
```
Comment (TejasQ) [inline src/Page/Page.tsx]: Out of curiosity, why are we renaming this?

Comment (mjopson) [reply src/Page/Page.tsx]: It used to always contain actions, but now it contains whatever is on "the end", which might not be actions anymore.

Comment (fabien0102) [reply src/Page/Page.tsx]: For me, this container should be semantically `ActionContainer`, after we "just" need to move this `ActionContainer` to `start` or `end` regarding the options.

I think that we can clean up a bit this implementation :wink:

**Comment by micha-f** [comment]
I think it is nice to have a bit more spacing when actions are on the left as that communicates that the page title does not belong to the actions. Looks good.

**Comment by mjopson** [comment]
An issue is that ActionsContainer has a left margin on it due to it being the second item.  When the order is reversed, the margin creates a problem.

Perhaps this instead: we have TitleBar reorder and add margins to its children.  We use right margin rather than left for simplicity, because we can safely add it to all of the elements rather than having to exclude the first (this would continue to work if more elements were added to the title bar).

In TitleBar:

  "& > :nth-child(n)": {
    marginRight: theme.space.element,
  },
  "& > :nth-child(2)": {
    order: actionsFirst ? -1 : undefined,
  }

"nth-child(n)" is a specificity hack - Title has defines its own margins that we are overriding.  The nth-child(2) here feels fragile; what I really want to say is "the ActionsContainer element".  Supposedly we can do this with syntax like the following:

[ActionsContainer]: {
  ....
}

But it seems this is only possible with babel-plugin-emotion.

**Comment by TejasQ** [comment]
I understand the concern, @mjopson. We actually need this feature rather quickly and so @fabien0102 from our team [was able to solve it already](https://github.com/contiamo/operational-ui/pull/683/files#diff-88ff81fac534225deda44f519755137c). I have to say – he has done a really good job as it does not really involve a lot of convolutions: it simply reverses flex direction and deals with the margin based on the passed in prop. Ideally, we should not have to override _our own_ CSS with specificity as we develop our own components. I'm not saying that we don't do this: I'm sure such hacks exist in this codebase – but rather, I'm suggesting we keep it to a minimum.

For this reason, I'm going to have to close this PR in favor of #683. I hope this is okay.

You're more than welcome to find another issue and submit a pull request. In fact, I'd encourage it. 😄  It may be more efficient to ask to take over an issue beforehand so we can evaluate its urgency and factor in the questions of _"do we need this sooner? Is this something the internal team should handle because it requires more rapid offline conversation and iteration?" etc.

Thanks again!

**Pull request closed without merging**